### PR TITLE
Fix and improve `KeyMap::get_keymap_descriptions()`

### LINKED ISSUE #843: open-all-unread-in-browser-and-mark-read is not correctly showing up in help screen of articlelist
**Newsboat version**: newsboat 2.19.0 (Arch Linux package `newsboat 2.19-1`)

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
bind-key S save-all articlelist
bind-key O open-all-unread-in-browser-and-mark-read articlelist
```

Steps to reproduce the issue:

1. Add following line to the config file:
    `bind-key O open-all-unread-in-browser-and-mark-read articlelist`

2. Start Newsboat and navigate into a feed (showing a list of articles)

3. Press `?` to open the help

**Expected**:
1. A line is shown containing `O     open-all-unread-in-browser-and-mark-read`
2.  `open-all-unread-in-browser-and-mark-read` is **not** shown below `Unbound functions`

**Actual**:
1. A line is shown containing `O`
2. `open-all-unread-in-browser-and-mark-read` is shown below `Unbound functions`

**Other info you think is relevant**:
image

### PR DESCRIPTION
**Pull request by dennisschagt**
Fixes #843 

I squashed several refactoring commits into one.
If you are interested, the separate small commits can still be found on a temporary branch: https://github.com/newsboat/newsboat/compare/master...dennisschagt:bugfix/issue-843-before-rebase

### CODE REVIEW

File: src/helpformaction.cpp [hunk 1]
```diff
@@ -74,8 +74,7 @@ void HelpFormAction::prepare()
 			fmt.do_format(cfg->get_configvalue("help-title-format"),
 				width));
 
-		const auto descs = v->get_keymap()->get_keymap_descriptions(
-				v->get_keymap()->get_flag_from_context(context));
```
Comment (Minoru) [inline src/helpformaction.cpp]: After this change, do we even need `get_flag_from_context()` anymore?

**Comment by Minoru** [comment]
I wanted to check if you had a 100% test coverage on `get_keymap_descriptions()` before you refactored it, so I checked out 34d3ed33, but the tests fail for me:

```
-------------------------------------------------------------------------------
get_keymap_descriptions() returns at most one entry per key
-------------------------------------------------------------------------------
test/keymap.cpp:226
...............................................................................

test/keymap.cpp:244: FAILED:
  REQUIRE( keys.count(key) == 0 )
with expansion:
  1 == 0
with message:
  key: "ENTER"

-------------------------------------------------------------------------------
get_keymap_descriptions() does not return empty commands or descriptions
-------------------------------------------------------------------------------
test/keymap.cpp:250
...............................................................................

test/keymap.cpp:263: FAILED:
  REQUIRE( description.cmd != "" )
with expansion:
  "" != ""

-------------------------------------------------------------------------------
get_keymap_descriptions() includes entries which include different keys bound
to same operation
-------------------------------------------------------------------------------
test/keymap.cpp:268
...............................................................................

test/keymap.cpp:284: FAILED:
  REQUIRE( description.key != "" )
with expansion:
  "" != ""
```

Did you accidentally broke them by rearranging commits during rebasing? (I've certainly fell into that trap before).

File: src/helpformaction.cpp (hunk 1 above)
Comment (dennisschagt) [reply src/helpformaction.cpp]: Yes, it is still used in the `get_keymap_description()` function itself: https://github.com/newsboat/newsboat/blob/606b84b0e216094efaa80d24c49ca17f594a59e2/src/keymap.cpp#L518

The flag is needed to determine if an operation is supported in the specified context.

Comment (Minoru) [reply src/helpformaction.cpp]: Oh. I'm still reviewing that part; should've held the comments off until then. Thanks for the explanation! :+1:

**Comment by dennisschagt** [comment]
> Did you accidentally broke them by rearranging commits during rebasing? (I've certainly fell into that trap before).

No, those tests show behavior which was already broken.
(I wrote most of them before fixing the issues).

### get_keymap_descriptions() returns at most one entry per key
```
test/keymap.cpp:226
...............................................................................

test/keymap.cpp:244: FAILED:
  REQUIRE( keys.count(key) == 0 )
with expansion:
  1 == 0
with message:
  key: "ENTER"
```
This shows that `get_keymap_descriptions()` was returning bindings multiple times (once for every context in which it is registered).
The only reason this wasn't visible is because the [`HelpFormAction` filtered out bindings with a different context](https://github.com/newsboat/newsboat/blob/606b84b0e216094efaa80d24c49ca17f594a59e2/src/helpformaction.cpp#L114-L116).

### get_keymap_descriptions() does not return empty commands or descriptions
```
test/keymap.cpp:250
...............................................................................

test/keymap.cpp:263: FAILED:
  REQUIRE( description.cmd != "" )
with expansion:
  "" != ""
```
For some reason, a second binding for the same operation (with a different key) would be included in the output but [without setting `.cmd` and `.help_text`](https://github.com/newsboat/newsboat/blob/6dd942a27f0bb71d0979b9b713e331b941a366b3/src/keymap.cpp#L534-L543).
This can be seen in screenshot at #843 (`O`-key on a line of itself).

### get_keymap_descriptions() includes entries which include different keys bound to same operation
```
test/keymap.cpp:268
...............................................................................

test/keymap.cpp:284: FAILED:
  REQUIRE( description.key != "" )
with expansion:
  "" != ""
```
This specific `REQUIRE` I squashed into the testcase (maybe better if I change it into a non-fatal `CHECK`).
Without it, the following `REQUIRE` would fail showing something like `with expansion: {"", "a"} == {"a", "b", "c"}`

This test shows that originally, operations without a default key were put in the keymap with an empty key (`key == ""`).
That caused the function to always be included as an unbound key in the help screen
(see bottom of screenshot at #843 where `open-all-unread-in-browser-and-mark-read` shows up even if the `O` key is bound to it).

**Comment by dennisschagt** [comment]
Just to summarize, I added tests which originally failed before committing the changes which fixed the behavior.

@Minoru: Is this fine with you?
Or, do you want me to commit tests after fixing the issue which would make them fail?

**Comment by Minoru** [comment]
Thanks for the explanations! Now it all makes sense :)

I think it's great that you've added a failing test before fixing its cause (TDD!), I just wish that:
- there were tests that document existing behaviour *before* the refactoring (to make sure refactor doesn't break anything);
- the commit that adds failing tests mentioned that they fail;
- bug fixes weren't mixed with refactors, if that's not too inconvenient (doesn't seem to be the case here, though — looks like it's easier to untangle the mess than to amend it). Preferably, add a test and the fix in a single commit, with a message that says "Fix …".

Can you add tests that demonstrate how this method behaved before the refactor? I'm ok with a force-push for that.

**Comment by dennisschagt** [comment]
> and the new code looks good to me.

Thanks!

> I would still appreciate some tests for pre-refactor behaviour, just to be on the safe side. Let me know if you're adding them

I'm working on those tests now.

**Comment by dennisschagt** [comment]
See following for diff between code as you reviewed it and the current version: https://github.com/newsboat/newsboat/compare/606b84b0e216094efaa80d24c49ca17f594a59e2..168e76ba4a038039a5477bff473c2ac53c77aa55

> I just wish that: there were tests that document existing behaviour before the refactoring

Done

> I just wish that: the commit that adds failing tests mentioned that they fail;

The tests which were failing before are now included in the commit which also fixes their respective issue (so none of the intermediate commits result in failing tests)

> I just wish that: bug fixes weren't mixed with refactors, <…> Preferably, add a test and the fix in a single commit, with a message that says "Fix …".

Done

**Comment by Minoru** [comment]
Just realised that 35d6ff5fd8d5cd95b682ad4eea9528e83d2b0dad might be seen as a breaking change. Previously, if multiple keys were bound to a single operation, help dialog displayed that in a more aesthetically pleasing way, without repeating the op name and description. However, it might've created confusion because it's not clear which operation the key belongs to — the previous one, or the next one? E.g.:

```
o    open    Open in browser
M
q    quit    Quit
```

Does <kbd>M</kbd> open in browser, or does it quit?

35d6ff5fd8d5cd95b682ad4eea9528e83d2b0dad removes the ambiguity by always listing the name and the description, so let's keep it.

@dennisschagt, thanks for the fixes and for literally re-doing the PR just to make it easier to review :heart:

**Pull request merged**